Create a `MessageDialog` API

SaschaCowley opened the pull request:
### Link to issue number:

Closes #13007
Closes #12344 
Closes #12353 

### Summary of the issue:

Currently, there are several ways to create message boxes/dialogs in NVDA, all of them with their own drawbacks.
Additionally, subclassing `wx.Dialog` (or using it directly) brings with it the risk of breaking parts of NVDA. See #13007 for in-depth discussion.

Thus, a new API for creating highly customizable dialogs that work well with NVDA is desired.

### Description of user facing changes

This PR does not, in itself, introduce any end-user facing changes. However, it lays the groundwork for closing a number of issues.

### Description of development approach

Subclassed `wx.Dialog` to add the desired functionality in #13007

Re-implemented the about box to use a `MessageDialog`.

Added a callback field to `gui.blockAction._Context` to allow adding a function to execute when the block condition is met.

### Features and tasks

* [x] Rename buttons
* [x] Add custom buttons
* [x] Associate context help with the given message box
* [x] Show the dialog as a modal or a non-modal
* [x] for non modal dialogs: 
  * [x] provide a default return code when dialogs are forced to close (eg via an exit, perform cancel) (WIP)
  * [x] attach a callback to handle return codes
  * [x]  refocus a dialog in the event of needing to close it
  * [x] if a default return code is provided, make the dialog close-able with alt+F4 and escape (WIP)
* [x] for modal dialogs: 
  * [x] if wx.CANCEL is provided, make the dialog close-able with alt+F4 and escape (mostly done)
  * [x] nice to have:
    * [x] if wx.CANCEL is provided, perform this action when dialogs are forced to close.
    * [x] refocus a dialog in the event of needing to close it
* [x] Ensure that the test `Quits from keyboard with about dialog open` is re-enabled and works.
* [x] Block exiting NVDA with modal `MessageDialog`s open.
* [x] Focus open `MessageDialog`s without default actions when exiting.
* [x] Tests
  * [x] Unit tests
* [x] Document the new API in the developer guide
* [x] Migrate the new API from `gui.MessageDialog` to `gui.message`.
* [x] Mark other means of creating message dialogs as deprecated.
  * [x] Re implement their internals to use the new API
    * [x] `gui.message.messageBox`
    * [x] `gui.nvdaControls.MessageDialog`
    * [x] `gui.nvdaControls._ContinueCancelDialog`

### Testing strategy:

Unit and manual tests.

Unit tests are extensive, and reach around 85% coverage of the newly added code.

Message box shim tested with a variety of inputs and unittests. This is a fairly straightforward adapter.

`nvdaControls.MessageDialog` tested that screen curtain and add-on install warning dialogs still work as expected. Its private subclass, `_ContinueCancelDialog`, was tested by attempting to run the CRFT.

### Known issues with pull request:

Internal dialogs are yet to be migrated. This should be done gradually before the existing means of creating message dialogs are removed.

Slightly odd behaviour when a callback function blocks, and `focusBlockingInstances` is called, where a hidden dialog is focused. This can be worked around, but possibly not without unintended consequences.

Some invalid combinations (namely a button which is set as the default action but set not to close the dialog) are silently changed. While documenting this should be sufficient, logging or raising an exception may be warranted.

Visual proportions of the re-implemented About dialog are different.

The `ctrl+c` behaviour of copying the dialog's title, contents and buttons is not supported.

### Future work

* Remove all internal use of deprecated message dialog APIs.
* Remove deprecated message dialog APIs in 2026.1.

Nice to have

* Implement system tests for the new API.
* Comprehensively document a manual test plan, and develop any associated scripts.
* Add the ability to set a callback after a button has been added.
* Add a helper to add a "don't show this message again" checkbox.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes for NVDA 2025.1

- **New Features**
	- Support for math in PDFs, including MathML formulas.
	- New commands to adjust the volume of other applications.
	- Ability to navigate per sentence in Microsoft PowerPoint using keyboard shortcuts.
	- Enhanced URL reporting in Mozilla Firefox, including text fragments.
	- Added actions in the Add-on Store for managing add-on installations.
	- Support for specifying mirror URLs for updates and the Add-on Store.
	- Sorting options in the Add-on Store by publication date.
	- Automatic language switching with SAPI5 and Microsoft Speech Platform voices.
	- Configuration for braille navigation to speak current lines or paragraphs.
	- Improved reporting of selection updates in Microsoft Word.

- **Documentation**
	- Added a new section on the message dialog API in the developer guide, detailing usage and customization options.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SaschaCowley committed:
Working refactored message dialog based on extant message dialog and draft spec

SaschaCowley committed:
Improve positioning

SaschaCowley committed:
Explicitly focus default item

SaschaCowley committed:
Improved types and type hints

SaschaCowley committed:
Switched to setLabelText to avoid mnemonic issue

SaschaCowley committed:
Switched to icon bundles

SaschaCowley committed:
Renamed private methods to be dunder

SaschaCowley committed:
Some manual test cases

SaschaCowley committed:
Added check for default item before focusing

SaschaCowley committed:
Minor changes to make temp code work better

SaschaCowley committed:
Implementation of addButton, addOkButton and addCancelButton

SaschaCowley committed:
Return self for chaining, and add a option to set as default

SaschaCowley committed:
Added close functionality

SaschaCowley committed:
Fixed layout issues and updated test case

SaschaCowley committed:
Added post popup call

SaschaCowley committed:
Added infrastructure for updates to API not yet implemented

SaschaCowley committed:
Merge branch 'master' into messageDialogApi

SaschaCowley committed:
Swiutch to using partials for callback calling

SaschaCowley committed:
Added translator comments and some docstrings

SaschaCowley committed:
Added type checking to `__call_callback`

SaschaCowley committed:
Made `addButton` a `singledispatchmethod` that will add a `MessageDialogButton` if provided

SaschaCowley committed:
Refactored `addOkButton` and `addCancelButton` to be `partialmethod`s rather than normal functions.

SaschaCowley committed:
Added multi button add method

SaschaCowley committed:
Documentation improvements

SaschaCowley committed:
Added deque to track open MessageDialog instances

SaschaCowley committed:
Refactored to use a registry oif callbacks

SaschaCowley committed:
Added adders for oother default button types

SaschaCowley committed:
Added custom button IDs

SaschaCowley committed:
Various code clean-ups

SaschaCowley committed:
Partially working modal dialogs

SaschaCowley committed:
Better handling of modal dialogs (still not woroking properly as blockingInstancesExist tries to access destroyed instances)

SaschaCowley committed:
Fixed problem with blocking instances exist

SaschaCowley committed:
WIP better performance tracing and layout encapsulation

SaschaCowley committed:
Reorganised MessageDialog code

SaschaCowley committed:
Fixed return codes not working

SaschaCowley committed:
Explicitly hide dialog when closing

SaschaCowley committed:
Added escape code enum to support ID_NONE and ID_ANY

SaschaCowley committed:
Changed MessageDialogButton.default to MessageDialogButton.default_focus, and added a MessageDialogButton.default_action field.

SaschaCowley committed:
Corrected variable names

SaschaCowley committed:
Removed staticmethods to classmethods

SaschaCowley committed:
Made isBlocking a property

SaschaCowley committed:
Improved type annotations

SaschaCowley committed:
Documentation improvements

SaschaCowley committed:
Made Show and ShowModal work in line with wx.Dialog expectations

SaschaCowley committed:
Added help id argument

SaschaCowley committed:
Added exception for trying to show a MessageDialog without buttons

SaschaCowley committed:
Renamed DefaultMessageDialogButtons to DefaultMessageDialogButton

SaschaCowley committed:
Added an enumeration of common button combinations, and a buttons parameter to MessageDialog's initializer

SaschaCowley committed:
Used MessageDialog for about dialog

CyrilleB79 commented:
Thanks for addressing this issue.

@SaschaCowley, in the initial description of the issue, could you explicitly indicate that the current possibilities to create a message box / dialog in NVDA and their associated drawbacks are described in #13007? As well as the risk associated with sub-classing `wx.Dialog` or using it directly?

This way we exactly know that we just have to read #13007 to have the full context.

seanbudd commented:
General approach looks good to me, great work on this

SaschaCowley committed:
Added deprecation warning to gui.message.messageBox

SaschaCowley committed:
Re-enabled system test "Quits from keyboard with about dialog open"

SaschaCowley committed:
Updated block action to bring blocking dialogs to foreground

SaschaCowley committed:
Default action forces close

SaschaCowley committed:
Added a docstring

SaschaCowley committed:
Added a draft class-level docstring

SaschaCowley committed:
Respect CloseEvent.CanVeto

SaschaCowley committed:
Initial message dialog unit tests

SaschaCowley committed:
Refactored to use a common base class

SaschaCowley committed:
Removed default button assignment (reimplement later)

SaschaCowley committed:
Tests for adding standard buttons

SaschaCowley committed:
Added support for adding DefaultMessageDialogButtons with MessageDialog.addButtons without unpacking

SaschaCowley committed:
Added methods to easily add default buttons and unit tests for same

SaschaCowley committed:
Add hasDefaultAction property

SaschaCowley committed:
Add docstrings

SaschaCowley committed:
Improved implementation of default action getters

SaschaCowley committed:
Slightly refactored tests and improved docstrings for some tests

SaschaCowley committed:
Slight refacter to closing logic

josephsl commented:
Hi,

One thing that should be resolved: gui.nvdaControls includes a MessageDialog class. Do you plan to deprecate that one in favor of the proposed gui.messageDialog.MessageDialog class or merge what's avaiallbe?

Thanks.

SaschaCowley commented:
The current plan is to deprecate the existing `gui.NVDAControls.MessageDialog` in favour of this one. Exactly where the new `MessageDialog` class will be when released is not entirely clear yet, but I believe it will most likely be in `gui.message`. `gui.NVDAControls.MessageDialog` will most likely be removed in 2026.1.

SaschaCowley committed:
Re-implemented get default action and tests for same

SaschaCowley committed:
Made _getDefaultActionOrFallback comply with the return type opf _getDefaultAction

SaschaCowley committed:
Added additional setters and tests

SaschaCowley committed:
Call default callback when closing programatically

SaschaCowley committed:
Added docstrings to tests

SaschaCowley committed:
Made fallback default action more robust

SaschaCowley committed:
Renamed several double-underscored variables and methods to be single-underscored

SaschaCowley committed:
Added tests and docstrings

SaschaCowley committed:
Added lots of docstrings.

SaschaCowley committed:
Made a bunch of class names more usable

SaschaCowley committed:
Fixed some type info

SaschaCowley committed:
Added tests for _flatten_buttons

SaschaCowley committed:
Made the Show method comply with that of wx.Dialog

SaschaCowley committed:
Documentation improvements

SaschaCowley committed:
Restored behaviour of having an OK button as default

SaschaCowley committed:
Fixed label not being applied to buttons.

SaschaCowley committed:
Made id and button positional-only

SaschaCowley commented:
I'm not sure where the best place would be to put a primer on using these classes, if it's warranted at all. I've attached a sketch of some notes I've written as a start.
[messageDialog.txt](https://github.com/user-attachments/files/17643100/messagedialog.txt)

SaschaCowley commented:
@seanbudd I have marked this as ready for review not because I think it is perfect, but because I think external perspective is necessary. In particular, testing ideas would be appreciated. I do think that the code will be largely the same as what is merged, barring everybody despising the new API.

XLTechie commented:
I look forward to reading the full code of this, but I've just read the notes for now.
I'm very glad to have this brought forward!

Why is the developer guide not the obvious place for your notes?

I love the method chaining!

I would prefer that the "default action" be renamed to "fallback action" or something similar. Having a default action the main purpose of which is to fire on dialog close, and to be different than default focus, needs a less confusable name IMO.

gerald-hartig commented:
@SaschaCowley this is a great addition to the codebase. Very nice, well designed and comprehensive. 

I have some general notes for consideration - possibly for the next iteration of this.

Notes on the primer:

- It might be worth commenting on the benefits of method chaining to create more readable code. 
- The explanation of "default action" you could make more explicit: in terms of providing a safe and predictable way to handle unexpected dialog closures. 
- The primer doesn't explicitly state how button IDs are managed, especially when adding custom buttons. It should clarify whether developers need to manage IDs manually or if the system assigns them automatically.
- No information is provided on how to handle errors. Is it possible to set up invalid button configurations?
- The primer has simple examples but lacks a demonstration of a more advanced scenario, like using custom button labels, icons or handling dialog events beyond button clicks.
- The primer doesn't discuss how developers would discover the available DefaultButtonSet options or the ReturnCode values.

General notes for consideration:

- Introduce helper constructors or factory methods for simple common dialogs (e.g., MessageDialog.ok(message), MessageDialog.yesNo(message, callback)) to reduce boilerplate for simple scenarios.
- Replace EscapeCode.DEFAULT and EscapeCode.NONE with more descriptive enum values like DefaultAction.CANCEL_OR_OK and DefaultAction.DISABLE_CLOSE.

seanbudd commented on source/gui/__init__.py:
```diff
@@ -31,6 +31,7 @@
 	# be cautious when removing
```
general notes:

- there's a lot of strings that are not translatable which should be
- there's a lot of `snake_case` not `camelCase`
- can you ensure all usages of `isModalMessageBoxActive` get updated to focus the dialogs where appropriate

SaschaCowley replied:
I believe these have all been addressed.

seanbudd commented on source/gui/__init__.py:
```diff
@@ -367,7 +369,7 @@ def onInputGesturesCommand(self, evt):
 
 	def onAboutCommand(self, evt):
 		# Translators: The title of the dialog to show about info for NVDA.
-		messageBox(versionInfo.aboutMessage, _("About NVDA"), wx.OK)
```
I would encourage changing messageBox to use MessageDialog to ensure the API is compatible. Same with the other message dialogs

seanbudd commented on source/gui/__init__.py:
```diff
@@ -575,6 +577,29 @@ def onConfigProfilesCommand(self, evt):
 		ProfilesDialog(mainFrame).Show()
 		self.postPopup()
 
+	def onModelessOkCancelDialog(self, evt):
```
Is the plan to remove this? if not please mark it as private

seanbudd commented on source/gui/blockAction.py:
```diff
@@ -74,7 +75,15 @@ def _wrap(func):
 			for context in contexts:
 				if context.blockActionIf():
-					queueHandler.queueFunction(queueHandler.eventQueue, ui.message, context.translatedMessage)
+					if context == Context.MODAL_DIALOG_OPEN:
+						# Import late to avoid circular import
+						from gui.messageDialog import MessageDialog
+
+						if MessageDialog.BlockingInstancesExist():
+							MessageDialog.FocusBlockingInstances()
```
I think we should make a callback function part of the `Context` structure

seanbudd commented on source/gui/message.py:
```diff
@@ -97,6 +98,11 @@ def messageBox(
 	Because an answer is required to continue after a modal messageBox is opened,
 	some actions such as shutting down are prevented while NVDA is in a possibly uncertain state.
 	"""
+	warnings.warn(
+		DeprecationWarning(
+			"gui.message.messageBox is deprecated. Use gui.message.MessageDialog instead.",
+		),
+	)
```
can this warning be added to other message dialogs?

seanbudd commented on source/gui/messageDialog.py:
```diff
@@ -0,0 +1,722 @@
+	SAVE_NO_CANCEL = (
+		DefaultButton.SAVE,
+		# Translators: A don't save button on a message dialog.
+		DefaultButton.NO.value._replace(label=_("Do&n't save")),
```
I think it might be better to make this a different button

SaschaCowley replied:
This was chosen because of the value used in [`wx.StdDialogButtonSizer`](https://docs.wxpython.org/wx.StdDialogButtonSizer.html), the documentation for which says:

> If you add a button with `wx.ID_SAVE`, on macOS the button will be renamed to “Save” and the `wx.ID_NO` button will be renamed to “Don’t Save” in accordance with the macOS Human Interface Guidelines.

Which implies to me that `wx.ID_SAVE` and `wx.ID_NO` is the most appropriate pairing of IDs in this situation.

seanbudd commented on source/gui/messageDialog.py:
```diff
@@ -0,0 +1,722 @@
+		self.AddMainButtonId(buttonId)
+		# Default actions that do not close the dialog do not make sense.
+		if default_action and not closes_dialog:
+			log.warning(
+				"Default actions that do not close the dialog are not supported. Forcing close_dialog to True.",
+			)
```
should this be setting a variable too? I see it's done implicitly but perhaps it should be explicit

```suggestion
			log.warning(
				"Default actions that do not close the dialog are not supported. Forcing close_dialog to True.",
			)
			close_dialog = True
```

seanbudd commented on source/gui/messageDialog.py:
```diff
@@ -0,0 +1,722 @@
+			if dialog.isBlocking:
+				dialog.Raise()
+				dialog.SetFocus()
+				break
```
maybe this should happen for all instances? i.e. all get raised and maybe the last one gets focused?

seanbudd commented on source/gui/messageDialog.py:
```diff
@@ -0,0 +1,722 @@
+	# endregion
+
+	# region Methods for subclasses
+	def _addButtons(self, buttonHelper: guiHelper.ButtonHelper) -> None:
+		"""Adds additional buttons to the dialog, before any other buttons are added.
+		Subclasses may implement this method.
+		"""
+
+	def _addContents(self, contentsSizer: guiHelper.BoxSizerHelper) -> None:
+		"""Adds additional contents  to the dialog, before the buttons.
+		Subclasses may implement this method.
+		"""
```
these may have to be public if we expect add-ons to subclass them

SaschaCowley replied:
These names were chosen in line with those used in the old message dialog API. I'm happy to make them public, but we also need to flag that they must not be called externally. I could either delete them after they have been called, or raise an Exception if they're called externally, but I feel that that makes the API confusing (IDEs will not understand this convention, I think). I could also make them idempotent by replacing them with a noop after calling them in `__init__`, but again, does that go against developer expectations?

seanbudd replied:
I think its fine to leave as-is for now

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,431 @@
+
+
+class Test_MessageDialog_Buttons(MDTestBase):
+	def test_addOkButton(self):
+		"""Test adding an OK button to the dialog."""
+		self.dialog.addOkButton()
+		with self.subTest("Check button types"):
+			self.assertIsInstance(self.dialog.FindWindowById(wx.ID_OK), wx.Button)
+		with self.subTest("Test in main buttons"):
+			self.assertEqual(self.dialog.GetMainButtonIds(), [wx.ID_OK])
+		with self.subTest("Test has default action."):
+			self.assertTrue(self.dialog.hasDefaultAction)
+		with self.subTest("Test default action assignment."):
+			id, callback = self.dialog._getDefaultAction()
+			self.assertEqual(id, ReturnCode.OK)
+			self.assertIsNotNone(callback)
```
could these tests for adding buttons be abstracted / made DRY?

SaschaCowley replied:
Not easily with unittest alone. The [parameterized package](https://github.com/wolever/parameterized/) package seems to be a good drop-in means of adding parameterized tests to unittest though. This can also be done with a lot of fiddling with metaclasses, but I don't want to add code that needs to be tested to tests.
https://github.com/nvaccess/nvda/blob/bd63423cb9486c0829138fe65e8608fb225ff84f/tests/unit/test_messageDialog.py#L90-L153

seanbudd reviewed (commented):
Approach generally LGTM some general comments

josephsl commented:
Hi all,

A few testing results (version: pr17304-34507,e71d3c79)

1. NVDA menu/dialog submenu:

Contrary to expectations, when clicking items in NVDA menu/Dialog submenu, nothing pops up. For example, clicking "OK" menu item does nothing whereas code specifies opening a modal dialog. Because the dialog submenu is shown in binary builds, I think it would be best to hide it unless explicitly requested while testing.

2. Main thread check failure when add-on onInstall function shows a message dialog (possibly accompanied by raising an exception):

Some add-ons (including Windows App Essentials) displays error messages if NVDA or system requiremnt is not met for the ad-on. Some add-ons go further and raise an exception, causing add-on handler to terminate add-on installation. When coding the error dialog using the proposed MessageDialog API with ShowModal call, wxPython raises main thread check failure (mostly dealing with timer) and keyboard no longer responds (the only way out is logging out via Windows sign-in dialog (Control+Alt+Delete)). This might be a fault with add-on developer (in this context ,me) or this behavior might have been anticipated (I would imagine people would just substitute gui.messageBox with gui.MessageDialog API call). What's more interesting is that the add-on error dialog doesn't even show up when called using just MessageDialog(args).Show() method but that doesn't take down the keyboard.

Thanks.

SaschaCowley commented:
@josephsl 

>     1. NVDA menu/dialog submenu:
> 
> Contrary to expectations, when clicking items in NVDA menu/Dialog submenu, nothing pops up. For example, clicking "OK" menu item does nothing whereas code specifies opening a modal dialog. Because the dialog submenu is shown in binary builds, I think it would be best to hide it unless explicitly requested while testing.

This is simply because I haven't implemented all of these functions, as they're just there for testing. They will be removed before this is merged.

>     2. Main thread check failure when add-on onInstall function shows a message dialog (possibly accompanied by raising an exception):
> 
> Some add-ons (including Windows App Essentials) displays error messages if NVDA or system requiremnt is not met for the ad-on. Some add-ons go further and raise an exception, causing add-on handler to terminate add-on installation. When coding the error dialog using the proposed MessageDialog API with ShowModal call, wxPython raises main thread check failure (mostly dealing with timer) and keyboard no longer responds (the only way out is logging out via Windows sign-in dialog (Control+Alt+Delete)). This might be a fault with add-on developer (in this context ,me) or this behavior might have been anticipated (I would imagine people would just substitute gui.messageBox with gui.MessageDialog API call). What's more interesting is that the add-on error dialog doesn't even show up when called using just MessageDialog(args).Show() method but that doesn't take down the keyboard.

Thanks for pointing this out, I'll have to investigate further. I'll let you know if I need more info for debugging this.

SaschaCowley commented:
@XLTechie 

> Why is the developer guide not the obvious place for your notes?

It probably is, no section of the guide jumped out at me as being the best place to put the notes though.

> I love the method chaining!

Glad to hear it! It's one of the few features I regularly miss from Java when working in Python.

SaschaCowley commented:
@XLTechie

> I would prefer that the "default action" be renamed to "fallback action" or something similar. Having a default action the main purpose of which is to fire on dialog close, and to be different than default focus, needs a less confusable name IMO.

@gerald-hartig 

> - The explanation of "default action" you could make more explicit: in terms of providing a safe and predictable way to handle unexpected dialog closures.

Both great points. I will rename this and provide a more detailed description in the developer guide.

SaschaCowley commented:
@gerald-hartig 

Thanks for the detailed feedback. I'll take on all of these suggestions when working on the developer documentation for this API.

>     * Introduce helper constructors or factory methods for simple common dialogs (e.g., MessageDialog.ok(message), MessageDialog.yesNo(message, callback)) to reduce boilerplate for simple scenarios.

Great idea, I'll have a look at doing this.

>     * Replace EscapeCode.DEFAULT and EscapeCode.NONE with more descriptive enum values like DefaultAction.CANCEL_OR_OK and DefaultAction.DISABLE_CLOSE.

These names are based on the WX equivalents, which are `wx.ID_ANY` and `wx.ID_NONE`. I will definitely consider more descriptive names for them though!

SaschaCowley committed:
Rename several button properties to bring them into line with NVDA coding style and make the code more self-documenting.

* Rename `default_focus` to `defaultFocus`
* Rename `default_action` to `fallbackAction`
* Rename `closes_dialog` to `closesDialog`

SaschaCowley committed:
Parameterized tests for the default button adders to make code more DRY

XLTechie commented:
@SaschaCowley

> It probably is, no section of the guide jumped out at me as being the best place to put the notes though.

That's why I created a new one for the Extension Points chapter.
I suggest you do the same thing here.

A new chapter called "Communicating with the user" or "Communicating to the user" or similar.
We can talk about ui.message and ui.browseableMessage in the future, and the speech/braille specific versions. For now, it can just be your dialog API.

> Glad to hear it! It's one of the few features I regularly miss from Java when working in Python.

Ditto. It was one of the (few) things I really liked about Java, and was very disappointed to find it so rarely used in Python.

SaschaCowley committed:
Test that buttons are unique when adding

SaschaCowley committed:
Changed addButtons to accept an Collection of Buttons, rather than an unpacked tuple of Buttons and DefaultButtonSets

SaschaCowley committed:
Remove _flattenButtons

SaschaCowley committed:
Fixed some types

SaschaCowley committed:
Tightened some logic and made the close box enabled only when there is a valid EscapeId

SaschaCowley committed:
Fixed test dialog

SaschaCowley committed:
Formatted

SaschaCowley committed:
Reimplemented gui.message.messageBox using MessageDialog

SaschaCowley committed:
Re-implemented the message box shim

SaschaCowley committed:
Refactored and documented.

SaschaCowley commented:
@josephsl I believe the exception you encountered in https://github.com/nvaccess/nvda/pull/17304#issuecomment-2475732992 should be resolved now. Would you mind testing again with at least https://github.com/nvaccess/nvda/pull/17304/commits/fca7eb8244f10d3d574e4eec6d75f70c1e92bfff?

Regarding the errors encountered when calling `MessageDialog.ShowModal` from a non-main thread, this is expected behaviour. `wx.CallAfter` should be used instead. Do you, @seanbudd , or others, think that this should detect if being called in a non-main thread and correct for that?

josephsl commented:
Hi,

The error is resolved.

As for main thread detection: perhaps we could use main thread detection. However, it might be possible that an operation is performed right after dismissing this dialog that might block the main thread i.e. the dialog asks to perform a task that could take a long time to do such as network connectivity.

Thanks.

josephsl commented:
Hi,

As a follow-up: the error still happens if gui.messageDialog.MessageDialog.ShowModal() is called directly as proposed in the message dialog notes posted a few days ago, so it might be worth looking into main GUI thread detection. Another way could be reminding folks to wrap ShowModal call inside wx.CallAfter function.

Thanks for considering our feedback in implementingg this PR.

SaschaCowley committed:
Added a helper function for calling wx functions on the main thread and waiting for their return.

SaschaCowley committed:
Used new wxCallOnMain function foor message box shim

SaschaCowley committed:
Tidied up message box shim

SaschaCowley committed:
Added new function to changes

SaschaCowley committed:
Only allow instantiating message dialogs from the GUI thread

SaschaCowley committed:
Fixed implementation of isBlocking and FocusBlockingInstances

SaschaCowley committed:
Added helper methods for alert, confirm and query dialogs

SaschaCowley committed:
Added some methods to be implemented

SaschaCowley committed:
Added more main thread checks

SaschaCowley committed:
Started implementation of message.MessageDialog -> messageDialog.MessageDialog shim (untested)

SaschaCowley committed:
Fixed duplicate superclass of gui.nvdaControls._ContinueCancelDialog

SaschaCowley committed:
Refactored show checks into their own private method

SaschaCowley committed:
Old message dialog shimmed to new one and working with screen curtain (probably more to do still)

SaschaCowley committed:
Refactor checks to make them easier to override in subclasses.

SaschaCowley committed:
Shim for gui.nvdaControls.MessageDialog seems to work fine.

SaschaCowley committed:
Add sentinel value to allow discrimination of `None` and "not provided".

SaschaCowley committed:
Allow overriding the implicit return code of the action ID by providing a return code to MessageDialog.addButton or Button.

SaschaCowley committed:
Update shims of gui.nvdaControls.MessageDialog and gui.nvdaControls._ContinueCancelDialog to return the same wx constants as their previous implementations.

SaschaCowley committed:
Update changelog

SaschaCowley committed:
Added means of changing button labels.

SaschaCowley committed:
Added the ability to set the message text independent of the __init__

SaschaCowley committed:
Improve logic of alert, confirm and ask classmethods.

SaschaCowley committed:
Slight reorganisation in some methods.

SaschaCowley committed:
Code documentation improvements.

SaschaCowley committed:
Docstring for MessageDialog._commands

SaschaCowley committed:
Improved logic for whether dialogs are blocking

SaschaCowley committed:
Added callback field to gui.blockAction._Context, and refactored the dialog check to use same.

SaschaCowley committed:
Code style and documentation for gui.guiHelper.wxCallOnMain.

SaschaCowley committed:
Code clean up and documentation improvements to gui.message.messageBox.

SaschaCowley committed:
Documentation improvements to gui.nvdaControls.MessageDialog

SaschaCowley committed:
Removed old gui.nvdaControls.MessageDialog code.

SaschaCowley committed:
Added commit references to changes

SaschaCowley committed:
Fixed type hints for MethodCall

SaschaCowley committed:
Changed getDialogState to return a dictionary of button IDs and their labels

SaschaCowley committed:
Fixed return from getDialogState

SaschaCowley committed:
Added tests for setting button labels.

SaschaCowley committed:
Parameterised tests of setIcon and playSound to cover all dialog types.

SaschaCowley committed:
Fixed test name

SaschaCowley committed:
Added test for adding buttons with custom overrides.

SaschaCowley committed:
Removed use of deprecated log.warn

SaschaCowley committed:
Added test for adding buttons with non-unique IDs

SaschaCowley committed:
Fixed use of FindWindowById to FindWindow(id)

SaschaCowley committed:
Added tests for setting default focus.

SaschaCowley committed:
Removed unnecessary error case from _getFallbackAction

SaschaCowley committed:
Fixed logic for when the default focus is not in command registry

SaschaCowley committed:
Added further fallback action tests

SaschaCowley committed:
Added more tests for setting button labels

SaschaCowley committed:
Refactored MdTestBase into WxTestBase -> MdTestBase

SaschaCowley committed:
Added tests for threading

SaschaCowley committed:
Fixed test case name

SaschaCowley committed:
Added unittests for showing and hiding

SaschaCowley committed:
Improved modal show tests.

SaschaCowley committed:
Added tests for showEvent

SaschaCowley committed:
Fixed implementation of _getFallbackAction, _getFallbackActionOrFallback, and _executeCommand, and methods that call them, including unit tests.

SaschaCowley committed:
Added tests for blocking dialogs.

SaschaCowley committed:
Fixed issue when force-closing non-modal dialog

SaschaCowley committed:
Renamed _execute_command to _executeCommand

SaschaCowley committed:
Added tests for _onCloseEvent

SaschaCowley committed:
Added test for closeInstances

SaschaCowley committed:
Added tests for isBlocking

SaschaCowley committed:
Renamed hasDefaultAction to hasFallback to match rename of defaultAction to fallback

SaschaCowley committed:
Added tests for executeCommand

SaschaCowley committed:
Added guidance on dialog types

SaschaCowley committed:
Renamed setDefaultAction to setFallbackAction

SaschaCowley committed:
Doocstring improvements

SaschaCowley committed:
Renamed EscapeCode members to be more self documenting

SaschaCowley committed:
Renamed _realize_layout to _realizeLayout in line with NV Access style

SaschaCowley committed:
Added a bunch of missing docstrings

SaschaCowley committed:
Removed some commented out code

SaschaCowley committed:
Initial docs

SaschaCowley committed:
Added docstring to __init__

SaschaCowley committed:
Added documentation of convenience methods to dev docs

SaschaCowley committed:
Moved DialogType to gui.message

SaschaCowley committed:
Moved EscapeCode to gui.message

SaschaCowley committed:
Moved DialogType to gui.message

SaschaCowley committed:
Moved over the rest of gui.messageDialog to gui.message

SaschaCowley committed:
Updated copyright years

SaschaCowley committed:
Improvements to tests

SaschaCowley committed:
Renamed tests to be more in line with NVDA style

SaschaCowley committed:
Removed testing items from NVDA menu.

SaschaCowley committed:
Updated dev docs with information on importing

SaschaCowley committed:
Updated changes

SaschaCowley committed:
Deleted old messageDialog.py file

SaschaCowley committed:
Merge branch 'master' into messageDialogApi

# Conflicts:
#	user_docs/en/changes.md

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented:
Sorry this is such a big PR everyone.

@XLTechie I've added the documentation I've written to the developer guide as you suggested. @gerald-hartig the documentation I've written is now much more comprehensive. It should also format nicely when we get API documentation generation working.

@josephsl I believe I have fixed the bugs you were encountering. Note that most of this code is not thread safe, so simply using `MessageDialog`s from add-on code is not safe. I've added a function, `gui.guiHelper.wxCallOnMain` to synchronously make calls on the GUI thread and get the return or exceptions from them.

@gerald-hartig, @seanbudd, @XLTechie and @CyrilleB79 I've renamed several things to have less similar names and to be more self-documenting at all of your suggestions.

Sorry if I've missed anyone or mixed up who commented what; it's late and it's been a long couple of weeks. Thanks for all of your reviews so far, and I look forward to your further feedback.

josephsl commented:
Hi, thanks for all your work over the last few weeks. I think a compromise could be calling gui.MessageDialog.convenience (such as alert) class methods which will call the newly defined function anyway to keep the code more readable (inside add-ons and in core). There are other dialog types encountered such as error dialog, but I think that could be addressed in 2025.2 or later. Again thank you for all your hard work and taking our feedback.

seanbudd commented on requirements.txt:
```diff
@@ -27,6 +27,7 @@ nuitka==2.5.4
 
 # Creating XML unit test reports
 unittest-xml-reporting==3.2.0
+parameterized==0.9.0 
```
Can this get a comment?

```suggestion
# Feed parameters to tests neatly
parameterized==0.9.0 
```

seanbudd commented on source/gui/blockAction.py:
```diff
@@ -1,25 +1,42 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+from collections.abc import Callable
```
why was this changed from the `typing` import?

SaschaCowley replied:
Since Python 3.9, [`typing.Callable` is a deprecated alias to `collections.abc.Callable`](https://docs.python.org/3/library/typing.html#typing.Callable).

seanbudd commented on source/gui/blockAction.py:
```diff
@@ -1,25 +1,42 @@
+import core
+
+
+def _isModalMessageBoxActive() -> bool:
+	from gui.message import isModalMessageBoxActive
```
```suggestion
def _isModalMessageBoxActive() -> bool:
	"""Avoid circular import of isModalMessageBoxActive"""
	from gui.message import isModalMessageBoxActive
```

seanbudd commented on source/gui/guiHelper.py:
```diff
@@ -458,3 +462,55 @@ class SIPABCMeta(wx.siplib.wrappertype, ABCMeta):
+class _WxCallOnMainResult:
+	"""Container to hold either the return value or exception raised by a function."""
+
+	__slots__ = ("result", "exception")
```
TIL about slots!
Can you add[ type annotations?](https://stackoverflow.com/a/52553217)

```suggestion
	__slots__ = ("result", "exception")
	
	result: Any
	exception: Exception
```

seanbudd commented on source/gui/guiHelper.py:
```diff
@@ -458,3 +462,55 @@ class SIPABCMeta(wx.siplib.wrappertype, ABCMeta):
+) -> _WxCallOnMain_T:
+	"""Call a non-thread-safe wx function in a thread-safe way.
+
+	Using this function is prefferable over calling :fun:`wx.CallAfter` directly when you care about the return time or return value of the function.
```
```suggestion
	Using this function is preferable over calling :fun:`wx.CallAfter` directly when you care about the return time or return value of the function.
```

seanbudd commented on source/gui/guiHelper.py:
```diff
@@ -458,3 +462,55 @@ class SIPABCMeta(wx.siplib.wrappertype, ABCMeta):
+	:raises Exception: If `function` raises an exception, it is transparently re-raised so it can be handled on the calling thread.
+	:return: Return value from calling `function` with the given positional and keyword arguments.
+	"""
+	result = _WxCallOnMainResult()
+	event = threading.Event()
+
+	def functionWrapper():
+		try:
+			result.result = function(*args, **kwargs)
+		except Exception:
+			result.exception = sys.exception
+		event.set()
+
+	if wx.IsMainThread():
+		functionWrapper()
+	else:
+		wx.CallAfter(functionWrapper)
+		event.wait()
+
+	try:
+		return result.result
+	except AttributeError:
+		# If result is undefined, exception must be defined.
+		raise result.exception  # type: ignore
```
This might be simpler and easier to understand without knowledge of slots.
Feel free to reject though.

```suggestion
	result: Any = None
	exception: Exception | None = None
	event = threading.Event()

	def functionWrapper():
		try:
			nonlocal result
			result = function(*args, **kwargs)
		except Exception:
			nonlocal exception
			exception = sys.exception
		event.set()

	if wx.IsMainThread():
		functionWrapper()
	else:
		wx.CallAfter(functionWrapper)
		event.wait()

	if exception is not None:
		raise exception
	return result
```

SaschaCowley replied:
TIL about `nonlocal`! Thanks, I like that a lot.

seanbudd commented on source/gui/guiHelper.py:
```diff
@@ -458,3 +462,55 @@ class SIPABCMeta(wx.siplib.wrappertype, ABCMeta):
+	*args: _WxCallOnMain_P.args,
+	**kwargs: _WxCallOnMain_P.kwargs,
+) -> _WxCallOnMain_T:
+	"""Call a non-thread-safe wx function in a thread-safe way.
```
Might be worth emphasising the blocking nature of this earlier.

```suggestion
	"""Call a non-thread-safe wx function in a thread-safe way.
	Blocks current thread.
```

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -270,120 +270,64 @@ def __init__(self, *args, **kwargs):
-			# Translators: A cancel button on a message dialog.
-			label=_("Cancel"),
+	@staticmethod
+	def _legasyDialogTypeToDialogType(dialogType: int) -> gui.message.DialogType:
```
```suggestion
	def _legacyDialogTypeToDialogType(dialogType: int) -> gui.message.DialogType:
```

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -110,6 +110,9 @@ Add-ons will need to be re-tested and have their manifest updated.
   It can be used in scripts to report the result when a boolean is toggled in `config.conf`
 * Removed the requirement to indent function parameter lists by two tabs from NVDA's Coding Standards, to be compatible with modern automatic linting. (#17126, @XLTechie)
 * Added the [VS Code workspace configuration for NVDA](https://nvaccess.org/nvaccess/vscode-nvda) as a git submodule. (#17003)
+* A new function, `gui.guiHelper.wxCallOnMain`, has been added, which allows safely and syncronously calling wx functions from non-GUI threads, and getting their return value. (#17304)
```
```suggestion
* A new function, `gui.guiHelper.wxCallOnMain`, has been added, which allows safely and synchronously calling wx functions from non-GUI threads, and getting their return value. (#17304)
```

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -414,29 +358,24 @@ def __init__(
 		if helpId is not None:
 			self.helpId = helpId
 		super().__init__(parent, title, message, dialogType)
+		if helpId is not None:
+			# Help event has already been bound (in supersuperclass), so we need to re-bind it.
+			self.bindHelpEvent(helpId, self)
```
this is a little concerning, doesn't that mean add-ons using the old `MessageDialog` need to do this otherwise face breaking changes? can this be done from the old `MessageDialog`?

seanbudd commented:
@coderabbitai review

seanbudd commented on source/gui/__init__.py:
```diff
@@ -30,6 +30,7 @@
 	# messageBox is accessed through `gui.messageBox` as opposed to `gui.message.messageBox` throughout NVDA,
```
extra notes:

- we also need to deprecate and replace `runScriptModalDialog`. Because it creates a modal dialog that isn't tracked in `isModalMessageBoxActive`
- this PR also closes a bunch of issues listed in `13007`. Can these be added to the PR description so when this is merged they are also closed

seanbudd commented on source/gui/message.py:
```diff
@@ -1,16 +1,30 @@
 # -*- coding: UTF-8 -*-
```
extra notes:

- isModalMessageBoxActive should be deprecated and the logic moved to `BlockingInstancesExist`. The purposes of these functions are the same and we should avoid duplications. I would suggest deprecating `displayDialogAsModal` too in favour of `MessageDialog.showModal`
-

seanbudd replied:
Fair enough, thanks for the explanation

seanbudd reviewed (commented):
Note this is a preliminary review and the following files haven't been reviewed in detail:

- projectDocs/dev/developerGuide/developerGuide.md
- tests/unit/test_messageDialog.py
- source/gui/message.py

SaschaCowley committed:
Made delay before block message a constant

SaschaCowley committed:
Fixed misspelling

SaschaCowley committed:
Added explicit import to first example in developer guide for clarity

SaschaCowley committed:
Made thread safety warning more prominant

SaschaCowley committed:
Added note on custom button IDs

SaschaCowley commented on source/gui/guiHelper.py:
```diff
@@ -458,3 +462,55 @@ class SIPABCMeta(wx.siplib.wrappertype, ABCMeta):
+		try:
+			result.result = function(*args, **kwargs)
+		except Exception:
+			result.exception = sys.exception
+		event.set()
```
@coderabbitai this is incorrect. This exception needs to be re-raised with all context in tact on the calling thread. This is exactly what `sys.exception` is intended for.

SaschaCowley committed:
Fixed outdated docstring

SaschaCowley committed:
Added error checking to _onCloseEvent

SaschaCowley committed:
Renamed event handlers to be more consistant

SaschaCowley committed:
Update requirements.txt

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Update source/gui/blockAction.py

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Rewrote wxCallOnMain to use nonlocal rather than helper class

SaschaCowley committed:
Update source/gui/guiHelper.py

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Update source/gui/guiHelper.py

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Update user_docs/en/changes.md

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Made dialog destruction and deletion more robust

SaschaCowley committed:
Fixed incorrect call to warn

SaschaCowley committed:
Made runScriptModal safer, and marked it for removal

SaschaCowley committed:
Merge branch 'i13007' of https://github.com/nvaccess/nvda into messageDialogApi

seanbudd commented on source/visionEnhancementProviders/screenCurtain.py:
```diff
@@ -192,19 +192,19 @@ def _exitDialog(self, result: int):
 		self.EndModal(result)
 
-	def _onDialogActivated(self, evt):
+	def _onActivateEvent(self, evt):
```
can the event get a type hint?

seanbudd commented on source/visionEnhancementProviders/screenCurtain.py:
```diff
@@ -192,19 +192,19 @@ def _exitDialog(self, result: int):
 		self.noButton.SetFocus()
 
-	def _onShowEvt(self, evt):
+	def _onShowEvent(self, evt):
```
can the event get a type hint?

SaschaCowley committed:
Added type hints to _onActivateEvent and _onShowEvent for the screen curtain warning dialog

seanbudd commented on source/gui/message.py:
```diff
@@ -153,3 +151,1084 @@ def displayError(self, parentWindow: wx.Window):
+				if (action := self._getFallbackAction()) is not None:
+					return action
+			except KeyError:
+				log.debug("fallback action was not in commands. This indicates a logic error.")
```
```suggestion
				log.error("fallback action was not in commands. This indicates a logic error.")
```

seanbudd commented on source/gui/message.py:
```diff
@@ -153,3 +151,1084 @@ def displayError(self, parentWindow: wx.Window):
+		return DialogType.STANDARD
+
+
+def _MessageBoxButtonStylesToMessageDialogButtons(flags: int) -> tuple[Button, ...]:
```
```suggestion
def _messageBoxButtonStylesToMessageDialogButtons(flags: int) -> tuple[Button, ...]:
```

seanbudd reviewed (commented):
Thanks @SaschaCowley - this is nearly done. Only minor comments.
Still need to review developerGuide.md and the unit tests

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+
+### The message dialog API
+
+The message dialog API provides a flexible way of presenting interactive messages to the user. The messages are highly customisable, with options to change icons and sounds, button labels, return values, and close behaviour, as well as to attach your own callbacks.
```
```suggestion
The message dialog API provides a flexible way of presenting interactive messages to the user.
The messages are highly customisable, with options to change icons and sounds, button labels, return values, and close behaviour, as well as to attach your own callbacks.
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+
+MessageDialog(
+	mainFrame,
+	"Hello world!",
```
It's probably worth making all these strings in these examples translatable

```suggestion
	_("Hello world!"),
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
````diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+For non-modal dialogs, the easiest way to respond to the user pressing a button is via callback methods.
+
+```py
+def read_changelog():
````
Can these use camelCase to encourage add-on authors to follow our coding guidelines so that add-ons are easier to merge into core where necessary

```suggestion
def readChangelog():
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+).addButton(
+	ReturnCode.YES,
+	label="Yes",
+	default_focus=True,
```
```suggestion
	defaultFocus=True,
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+	callback=download_update
+).addButton(
+	ReturnCode.NO,
+	default_action=True,
```
```suggestion
	defaultAction=True,
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+	callback=remind_later
+).addButton(
+	ReturnCode.HELP,
+	closes_dialog=False,
```
```suggestion
	closesDialog=False,
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+
+#### Fallback actions
+
+It is worth interrogating the fallback action further.
```
I don't think this sentence adds anything

```suggestion
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+The fallback action can also be set to `EscapeCode.NO_FALLBACK` to disable closing the dialog like this entirely.
+If it is set to any other value, the value must be the id of a button to use as the default action.
+
+[^fn_defaultFallbackOk]: In actuality, the value of `dialog.GetAffirmativeId()` is used to find the button to use as fallback if using `EscapeCode.CANCEL_OR_AFFIRMATIVE` and there is no button with `id=ReturnCode.CANCEL`.
```
these references (e.g. `[^fn_defaultFallbackOk]`) don't appear to work in the built HTML from appveyor

seanbudd replied:
Simplifying

```suggestion
The value of `dialog.GetAffirmativeId()` is used to find the button to use as fallback if using `EscapeCode.CANCEL_OR_AFFIRMATIVE` and there is no button with `id=ReturnCode.CANCEL`.
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
````diff
@@ -1445,3 +1445,238 @@ Please see the `EventExtensionPoints` class documentation for more information,
+
+```py
+# To call
+some_function(arg1, arg2, kw1=value1, kw2=value2)
+# on the GUI thread:
+wxCallOnMain(some_function, arg1, arg2, kw=value1, kw2=value2)
````
```suggestion
someFunction(arg1, arg2, kw1=value1, kw2=value2)
# on the GUI thread:
wxCallOnMain(someFunction, arg1, arg2, kw=value1, kw2=value2)
```

SaschaCowley committed:
Apply suggestions from code review

Mostly stylistic fixes.

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Changed some methods to use lower camel case

SaschaCowley committed:
Improved logging

SaschaCowley committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Removed footnotes from the developer guide

seanbudd commented on source/gui/message.py:
```diff
@@ -153,3 +151,1084 @@ def displayError(self, parentWindow: wx.Window):
+	setHelpLabel = partialmethod(setButtonLabel, ReturnCode.HELP)
+	setHelpLabel.__doc__ = "Set the label of the help button in the dialog, if there is one."
+
+	def setOkCancelLabels(self, okLabel: str, cancelLabel: str) -> Self:
+		"""Set the labels of the ok and cancel buttons in the dialog, if they exist."
+
+		:param okLabel: New label for the ok button.
+		:param cancelLabel: New label for the cancel button.
+		:return: Updated instance for chaining.
+		"""
+		self._setButtonLabels((ReturnCode.OK, ReturnCode.CANCEL), (okLabel, cancelLabel))
+		return self
+
+	def setYesNoLabels(self, yesLabel: str, noLabel: str) -> Self:
+		"""Set the labels of the yes and no buttons in the dialog, if they exist."
+
+		:param yesLabel: New label for the yes button.
+		:param noLabel: New label for the no button.
+		:return: Updated instance for chaining.
+		"""
+		self._setButtonLabels((ReturnCode.YES, ReturnCode.NO), (yesLabel, noLabel))
+		return self
+
+	def setYesNoCancelLabels(self, yesLabel: str, noLabel: str, cancelLabel: str) -> Self:
+		"""Set the labels of the yes and no buttons in the dialog, if they exist."
+
+		:param yesLabel: New label for the yes button.
+		:param noLabel: New label for the no button.
+		:param cancelLabel: New label for the cancel button.
+		:return: Updated instance for chaining.
+		"""
+		self._setButtonLabels(
+			(ReturnCode.YES, ReturnCode.NO, ReturnCode.CANCEL),
+			(yesLabel, noLabel, cancelLabel),
+		)
```
I think it's better to leave as-is, thanks for clarifying

SaschaCowley committed:
Removed __del__ that was failing

SaschaCowley committed:
Corrections to dev guide

SaschaCowley committed:
Corrects to dev guide

SaschaCowley committed:
Merge branch 'i13007' of https://github.com/nvaccess/nvda into messageDialogApi

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+				tpe.submit(dlg.__init__, None, "Test").result()
+
+	def test_showOnNonMain(self):
+		# self.app = wx.App()
```
```suggestion
```

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+
+	def test_showModalOnNonMain(self):
+		"""Test that showing a MessageDialog modally on a non-GUI thread fails."""
+		# self.app = wx.App()
```
```suggestion
```

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+			mocked_messageBoxCounter.__iadd__.assert_called_once()
+			mocked_messageBoxCounter.__isub__.assert_called_once()
+
+		# raise Exception
```
```suggestion
```

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+
+	@parameterized.expand(
+		(
+			("closableCanCallClose", True, True, True),
+			("ClosableCannotCallClose", True, False, False),
+			("UnclosableCanCallClose", False, True, False),
+			("UnclosableCannotCallClose", False, False, False),
```
is there a way to parametize these with argument names to enhance readability? it's okay if not, just would be handy

e.g.
```suggestion
			("closableCanCallClose", closesDialog=True, canCallClose=True, expectedCloseCalled=True),
			("ClosableCannotCallClose", True, False, False),
			("UnclosableCanCallClose", False, True, False),
			("UnclosableCannotCallClose", False, False, False),
```

SaschaCowley replied:
Yep, I can just create `NamedTuple`s for these parameter lists.

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+@patch("gui.mainFrame")
+@patch.object(wx.Dialog, "ShowModal")
+class Test_MessageDialog_ShowModal(MDTestBase):
+	def test_showModalNoButtons(self, mocked_showModal: MagicMock, _):
```
could you clarify what the underscore mean in this context?

SaschaCowley replied:
It's just a conventional way of showing that we don't care about the mock object that is patching `mainFrame`.

SaschaCowley replied:
Would you like me to add a comment explaining that?

seanbudd replied:
all good - just wanted to learn. It's somewhat self-explanatory

CyrilleB79 replied:
I know that we are only in unit tests and that using `_` this way is a common practice.
Though, in projects where `_` is already used by `gettext`, it's not recommended to also use `_` for unused variables to avoid confusion.

That's really minor though, so feel free to ignore this comment if you wish.

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+	):
+		"""Test that blockingInstancesExist is correct in a number of situations."""
+		MessageDialog._instances.extend(instances)
+		print(MessageDialog._instances)
```
This should be a logging statement if its retained

```suggestion
```

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+				mocked_messageBoxCounter.__isub__.return_value
+			) = mocked_messageBoxCounter
+			self.dialog.ShowModal()
+			print(mocked_messageBoxCounter.mock_calls)
```
```suggestion
```

seanbudd commented on tests/unit/test_messageDialog.py:
```diff
@@ -0,0 +1,903 @@
+
+	@parameterized.expand(
+		(
+			("modalWithFallback", True, True, True),
```
same here re naming parameters

seanbudd commented on source/gui/message.py:
```diff
@@ -153,3 +151,1084 @@ def displayError(self, parentWindow: wx.Window):
+	"""
+
+
+class DefaultButton(Button, Enum):
```
Sure thing, we can easily change it later if requested

seanbudd reviewed (commented):
Thanks @SaschaCowley for all your excellent work here! Full review complete!

SaschaCowley committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Made several parameterized argument lists namedtuples for readability

SaschaCowley committed:
Merge branch 'master' into messageDialogApi

# Conflicts:
#	user_docs/en/changes.md

seanbudd reviewed (approved):
yahoo congrats on finishing this!

SaschaCowley committed:
Fixed up admonissions

SaschaCowley committed:
Added notes on thread safety

gerald-hartig commented:
Nice work @SaschaCowley
